tree/dataframe/src/RDFSnapshotHelpers.cxx

```diff
@@ -864,20 +868,41 @@ void ROOT::Internal::RDF::UntypedSnapshotRNTupleHelper::Initialize()
          outputDir = fOutputFile->mkdir(fDirName.c_str());
    }
 
-   fWriter = ROOT::RNTupleWriter::Append(std::move(model), fNTupleName, *outputDir, writeOptions);
+   fWriter = ROOT::Experimental::RNTupleParallelWriter::Append(std::move(model), fNTupleName, *outputDir, writeOptions);
+}
+
+void ROOT::Internal::RDF::UntypedSnapshotRNTupleHelper::InitTask(TTreeReader *, unsigned int slot)
+{
+   if (!fFillContexts[slot]) {
+      fFillContexts[slot] = fWriter->CreateFillContext();
+      fEntries[slot] = fFillContexts[slot]->GetModel().CreateBareEntry();
+   }
 }
 
-void ROOT::Internal::RDF::UntypedSnapshotRNTupleHelper::Exec(unsigned int /* slot */, const std::vector<void *> &values)
+void ROOT::Internal::RDF::UntypedSnapshotRNTupleHelper::Exec(unsigned int slot, const std::vector<void *> &values)
 {
+   auto &fillContext = fFillContexts[slot];
+   auto &outputEntry = fEntries[slot];
    assert(values.size() == fFieldTokens.size());
    for (decltype(values.size()) i = 0; i < values.size(); i++) {
-      fOutputEntry->BindRawPtr(fFieldTokens[i], values[i]);
+      outputEntry->BindRawPtr(fFieldTokens[i], values[i]);
    }
-   fWriter->Fill(*fOutputEntry);
+   fillContext->Fill(*outputEntry);
+}
+
+void ROOT::Internal::RDF::UntypedSnapshotRNTupleHelper::FinalizeTask(unsigned int slot)
+{
+   // In principle we would not need to flush a cluster here, but we want to benefit from parallelism for compression.
+   // NB: RNTupleFillContext::FlushCluster() is a nop if there is no new entry since the last flush.
+   fFillContexts[slot]->FlushCluster();
```

Comment:
With Snapshot to TTree, users can specify a value (expressed in number of entries) for the size of the output TTree clusters via [fAutoFlush](https://root.cern/doc/master/structROOT_1_1RDF_1_1RSnapshotOptions.html#a3a766c08900892477189398d61fdd214). I wonder how this line impacts this feature, which at the moment is not supported for the Snapshot to RNTuple, but I imagine it would be requested at some point.

Reply:
Yes, that's a good point. In RNTuple, we have options to express the target cluster size, but no (automatic) setting for the number of entries per cluster. Note that `UntypedSnapshotTTreeHelperMT` is also questionable in that regard because it writes / flushes output trees in `FinalizeTask`. So for RDataFrame snapshotting to TTree in MT mode, it's already more of an upper guarantee, but I guess many cluster will actually be smaller than the user wants them...

This probably requires further discussion - I opened #19784 and will resolve this thread for the moment.